Mage/src/main/java/mage/game/permanent/PermanentImpl.java

```diff
@@ -881,14 +881,25 @@ public int markDamage(int damageAmount, UUID sourceId, Game game, boolean preven
         return damage(damageAmount, sourceId, game, preventable, combat, true, null);
     }
 
+    private Ability getWitherOrInfectAbilityFromToken(PermanentToken token) {
+        Ability witherOrInfectAbility  = token.abilities.get(WitherAbility.getInstance().getId()).orElse(null);
+        if (witherOrInfectAbility == null) {
+            witherOrInfectAbility = token.abilities.get(InfectAbility.getInstance().getId()).orElse(null);
+        }
+        return witherOrInfectAbility;
+    }
+
     @Override
     public int applyDamage(Game game) {
         if (markedDamage == null) {
             return 0;
         }
         for (MarkedDamageInfo mdi : markedDamage) {
             Ability source = null;
-            if (mdi.sourceObject instanceof Permanent) {
+            if (mdi.sourceObject instanceof PermanentToken) {
+                source = getWitherOrInfectAbilityFromToken((PermanentToken)mdi.sourceObject);
+            }
+            if (mdi.sourceObject instanceof Permanent && source == null) {
                 source = ((Permanent) mdi.sourceObject).getSpellAbility();
```

Comment:
PermanentTokens do not have a spellAbility, so this would end up being null. This would cause the addCounters method to default to creating an event that had the _target_ as the source. This would cause the "controlled by current player" condition in Hapatra to always be false when tokens with wither deal damage.